File: app/web/src/ws/protocol.ts
export type ServerMessage =
  | { type: "ready" }
  | { type: "recordingStarted"; recordingId: string }
  | { type: "recordingStopped"; recordingId: string }
  | { type: "error"; message: string }
  | {
      type: "acknowledge";
      recordingId: string;
      frame: number;
      parameterChanged: boolean;
    };

export type ClientMessage =
  | {
      type: "startRecording";
      simulationId: string;
      particleCount: number;
      particleRadius: number;
      worldWidth: number;
      worldHeight: number;
      gravity: number;
      targetDensity: number;
      pressureMultiplier: number;
      viscosityStrength: number;
      smoothingRadius: number;
      interactionStrength: number;
      interactionRadius: number;
    }
  | { type: "stopRecording"; frame: number; simulationId: string }
  | { type: "step"; frame: number; dt: number }
  | { type: "setGravity"; frame: number; gravity: number }
  | { type: "setTargetDensity"; frame: number; targetDensity: number }
  | { type: "setPressureMultiplier"; frame: number; pressureMultiplier: number }
  | { type: "setViscosityStrength"; frame: number; viscosityStrength: number }
  | { type: "setSmoothingRadius"; frame: number; smoothingRadius: number }
  | {
      type: "setInteractionStrength";
      frame: number;
      interactionStrength: number;
    }
  | { type: "setInteractionRadius"; frame: number; interactionRadius: number }
  | { type: "enablePushInteraction"; frame: number; x: number; y: number }
  | { type: "enablePullInteraction"; frame: number; x: number; y: number }
  | { type: "disableInteraction"; frame: number };

File: app/websocket/src/protocol.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Deserialize)]
#[serde(tag = "type", rename_all = "camelCase")]
pub enum ClientMessage {
    #[serde(rename_all = "camelCase")]
    StartRecording {
        simulation_id: String,
        particle_count: usize,
        particle_radius: f32,
        world_width: f32,
        world_height: f32,
        gravity: f32,
        target_density: f32,
        pressure_multiplier: f32,
        viscosity_strength: f32,
        smoothing_radius: f32,
        interaction_strength: f32,
        interaction_radius: f32,
    },
    #[serde(rename_all = "camelCase")]
    StopRecording,
    Step {
        frame: usize,
        dt: f32,
    },
    SetGravity {
        frame: usize,
        gravity: f32,
    },
    #[serde(rename_all = "camelCase")]
    SetTargetDensity {
        frame: usize,
        target_density: f32,
    },
    #[serde(rename_all = "camelCase")]
    SetPressureMultiplier {
        frame: usize,
        pressure_multiplier: f32,
    },
    #[serde(rename_all = "camelCase")]
    SetViscosityStrength {
        frame: usize,
        viscosity_strength: f32,
    },
    #[serde(rename_all = "camelCase")]
    SetSmoothingRadius {
        frame: usize,
        smoothing_radius: f32,
    },
    #[serde(rename_all = "camelCase")]
    SetInteractionStrength {
        frame: usize,
        interaction_strength: f32,
    },
    #[serde(rename_all = "camelCase")]
    SetInteractionRadius {
        frame: usize,
        interaction_radius: f32,
    },
    EnablePushInteraction {
        frame: usize,
        x: f32,
        y: f32,
    },
    EnablePullInteraction {
        frame: usize,
        x: f32,
        y: f32,
    },
    DisableInteraction {
        frame: usize,
    },
}

#[derive(Debug, Serialize)]
#[serde(tag = "type", rename_all = "camelCase")]
pub enum ServerMessage {
    Ready,
    #[serde(rename_all = "camelCase")]
    StartedRecording {
        recording_id: String,
    },
    #[serde(rename_all = "camelCase")]
    StoppedRecording {
        recording_id: String,
    },
    #[serde(rename_all = "camelCase")]
    Acknowledge {
        recording_id: String,
        frame: usize,
        parameter_changed: bool,
    },
    Error {
        message: String,
    },
}
